### PR TITLE
Implement generate_signals() and calculate_position_size() along with its respective unit tests topic/#73

### PR DESCRIPTION
**Pull request by paramarap**
topic/#73

NOTE: **DO NOT EDIT TEXT BEFORE CREATING YOUR PULL REQUEST**

### Required quality checklist for pull requests
- [ ] Ready for Code Review & "`Ready For Review`" label added.
- [ ] All unit tests are passing.
- [ ] Code does not add any new logical errors. 
- [ ] Pylint's code analysis is above 8.0.
- [ ] Code's quality and documentation is up to point.
- [ ] GitHub Issue number is mentioned in the PR description.

### CODE REVIEW

**Comment by TheFJcurve** [comment]
@paramarap you have copied the logic of simple moving averages. Markovian Models has a different logic. You would have to implement that logic.

**Comment by paramarap** [comment]
> @paramarap you have copied the logic of simple moving averages. Markovian Models has a different logic. You would have to implement that logic.

Got it, will work on this today!

**Comment by paramarap** [comment]
This is the approach I took. I found a python library with a simple Gaussian HMM and took inspiration from its documentation (HERE: [https://hmmlearn.readthedocs.io/en/latest/tutorial.html#training-hmm-parameters-and-inferring-the-hidden-states](url)). 

I had to implement a small function to actually create, train, and return a trained model.

I believe this is a good foundation and can be scalable. I have also been messing around with the code to add a possible stop-loss feature as was discussed by @itsniranjan12.

File: uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py [hunk 1]
```diff
@@ -4,30 +4,147 @@
...
+        Train a Hidden Markov Model (HMM) using the given data.
+
+        :features_scaled: The input data (e.g., stock features) that has been standardized.
+        :return: A trained HMM model.
+        """
+        # Create an instance of the Gaussian Hidden Markov Model (HMM)
+        # n_components=3: Specifies that the HMM will have 3 hidden states (e.g., bullish, bearish, neutral)
+        # covariance_type="diag": Assumes that the features are independent, so the covariance matrix is diagonal
+        # n_iter=100: Sets the maximum number of iterations for the training process
+        hmm = GaussianHMM(n_components=3, covariance_type="diag", n_iter=100)
```
Comment (TheFJcurve) [inline uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py:54]: question (non-blocking): Why have we chosen Gaussian HMM? What are all the options available? And why is this the optimal model?

File: uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py [hunk 2]
```diff
@@ -4,30 +4,147 @@
...
+            # Get the column name for the stock's price
+            price_col = f"{ticker}_price"
+
+            # Check if the price column exists in the data
+            if price_col in current_data.columns:
+                # Get the stock prices
+                prices = current_data[price_col]
+
+                # Calculate log returns (a way to measure price changes)
+                log_returns = prices.pct_change().apply(lambda x: np.log(1 + x))
```
Comment (TheFJcurve) [inline uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py:82]: current_data will be a single row of data (containing current stock price for all the stocks). There would be no percent changes.

File: uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py [hunk 3]
```diff
@@ -4,30 +4,147 @@
...
+
+        # Combine all features into a single 2D array (rows = time steps, columns = stocks)
+        features = np.column_stack(features)
+
+        # Standardize the features (make them have a mean of 0 and a standard deviation of 1)
+        scaler = StandardScaler()
+        features_scaled = scaler.fit_transform(features)
+
+        # Train the HMM using the standardized features
+        hmm = self.__train_hmm(features_scaled)
```
Comment (TheFJcurve) [inline uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py:102]: As aforementioned, we have a single row of data (current time). Is this good to modify our hmm model on?

File: uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py [hunk 4]
```diff
@@ -4,30 +4,147 @@
...
+
+            # Map the state to a trading signal
+            if state == 0:
+                self._positions[ticker] = StockPosition.LONG  # Buy the stock
+            elif state == 1:
+                self._positions[ticker] = StockPosition.SHORT
+            else:
+                self._positions[ticker] = StockPosition.HOLD
 
     def calculate_position_size(self, ticker: str, price: float, portfolio_value: float) -> float:
```
Comment (TheFJcurve) [inline uwqsc_algorithmic_trading/src/algorithms/hidden_markov_model_impl.py:120]: According to HMM research, HMM works with individual distributions for position size (say normal distribution for Apple, student T distribution for Microsoft). This implementation of position size is a basic price calculation done for simple moving average. It would be more useful to apply inferences done by HMM to predict the best position. 

Note that we will also have to create the individual distributions.

**Review by TheFJcurve — CHANGES_REQUESTED** [review]
Added a few comments for you to go through. Note that I encourage using AI for understanding the code, but not to raise code changes.
